File: packages/julia/src/Seguranca/Ataques/AtaqueIVsDegenerados.jl
"""
Cifras às vezes têm "chaves fracas" ou "IVs fracos" - entradas específicas
(tudo zero, tudo 0xFF, padrões alternados) que produzem saída degenerada mesmo
quando a maioria das entradas se comporta bem. Testa especificamente esses
casos extremos, que testes com entrada aleatória raramente cobrem.
"""
struct AtaqueIVsDegenerados <: Ataque end

nome(::AtaqueIVsDegenerados) = "IVs degenerados (zero, 0xFF, alternado)"

function executar(a::AtaqueIVsDegenerados, alvo::AlvoCriptografico)
    chave = chave_de_teste(alvo)
    tam_iv = tamanho_iv(alvo)
    tamanho_bloco = 32

    crescente = UInt8[UInt8(i) for i in 0:(tam_iv-1)]

    casos = Pair{String,Vector{UInt8}}[
        "zero" => zeros(UInt8, tam_iv),
        "0xFF" => fill(0xff, tam_iv),
        "alternado 0xAA" => fill(0xaa, tam_iv),
        "alternado 0x55" => fill(0x55, tam_iv),
        "crescente" => crescente,
    ]

    problemas = Dict{String,Any}()
    for (nome_caso, iv) in casos
        ks = gerar_keystream_bruto(alvo, chave, iv, "enc", tamanho_bloco)

        metade = div(tamanho_bloco, 2)
        periodico = ks[1:metade] == ks[metade+1:2*metade]

        bytes_unicos = length(Set(ks))
        if periodico || bytes_unicos < tamanho_bloco * 0.5
            problemas[nome_caso] = (periodico=periodico, bytes_unicos=bytes_unicos)
        end
    end

    if !isempty(problemas)
        detalhe = join(
            [
                "$(nome_caso) (periódico=$(problemas[nome_caso].periodico ? "sim" : "não"), bytes únicos=$(problemas[nome_caso].bytes_unicos)/$tamanho_bloco)"
                for (nome_caso, _) in casos if haskey(problemas, nome_caso)
            ],
            "; ",
        )
        return ResultadoAtaque(nome(a), true, "alta", detalhe, problemas)
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Nenhum dos $(length(casos)) IVs degenerados testados produziu saída anômala",
    )
end

File: packages/dart/lib/Seguranca/Ataques/AtaqueIVsDegenerados.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Cifras às vezes têm "chaves fracas" ou "IVs fracos" - entradas específicas
/// (tudo zero, tudo 0xFF, padrões alternados) que produzem saída degenerada
/// mesmo quando a maioria das entradas se comporta bem. Testa especificamente
/// esses casos extremos, que testes com entrada aleatória raramente cobrem.
class AtaqueIVsDegenerados extends Ataque {
  @override
  String nome() => 'IVs degenerados (zero, 0xFF, alternado)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = alvo.chaveDeTeste();
    final tamanhoIv = alvo.tamanhoIv();
    const tamanhoBloco = 32;

    final casos = <String, Uint8List>{
      'zero': Uint8List(tamanhoIv),
      '0xFF': Uint8List(tamanhoIv)..fillRange(0, tamanhoIv, 0xff),
      'alternado 0xAA': Uint8List(tamanhoIv)..fillRange(0, tamanhoIv, 0xaa),
      'alternado 0x55': Uint8List(tamanhoIv)..fillRange(0, tamanhoIv, 0x55),
      'crescente': Uint8List.fromList(
        List<int>.generate(tamanhoIv, (i) => i),
      ),
    };

    final problemas = <String, Map<String, dynamic>>{};
    for (final entry in casos.entries) {
      final ks = alvo.gerarKeystreamBruto(
        chave,
        entry.value,
        'enc',
        tamanhoBloco,
      );

      const metade = tamanhoBloco ~/ 2;
      final periodico = bytesToHex(ks.sublist(0, metade)) ==
          bytesToHex(ks.sublist(metade, metade + metade));

      final bytesUnicos = ks.toSet().length;
      final poucaVariedade = bytesUnicos < tamanhoBloco * 0.5;

      if (periodico || poucaVariedade) {
        problemas[entry.key] = {
          'periodico': periodico,
          'bytes_unicos': bytesUnicos,
        };
      }
    }

    if (problemas.isNotEmpty) {
      final detalhe = problemas.entries
          .map((e) {
            final info = e.value;
            return '${e.key} (periódico='
                '${info['periodico'] == true ? 'sim' : 'não'}, '
                'bytes únicos=${info['bytes_unicos']}/$tamanhoBloco)';
          })
          .join('; ');
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.alta,
        detalhe,
        problemas,
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Nenhum dos ${casos.length} IVs degenerados testados produziu '
          'saída anômala',
    );
  }
}
